sklearn/impute/_knn.py

```diff
@@ -115,6 +117,14 @@ class KNNImputer(_BaseImputer):
            [8. , 8. , 7. ]])
     """
 
+    _parameter_constraints = {
+        **_BaseImputer._parameter_constraints,
+        "n_neighbors": [Interval(Integral, 1, None, closed="left")],
+        "weights": [StrOptions({"uniform", "distance"}), callable],
+        "metric": [StrOptions({"nan_euclidean"}), callable],
```

Comment:
If we want to be more general, since it's apparently already defined, we could use `_NAN_METRICS` in case more nan-supporting metrics are added later.

```suggestion
        "metric": [StrOptions(set(_NAN_METRICS )), callable],
```